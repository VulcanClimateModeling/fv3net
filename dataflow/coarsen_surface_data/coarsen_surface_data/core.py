from src import utils
import pandas as pd
import xarray as xr
from dask.delayed import delayed
from src.data import cubedsphere
import subprocess
import logging
import tempfile

combine_subtiles = delayed(cubedsphere.combine_subtiles)


@delayed
def concat_files(tiles):
    tile_nums = [int(tile) for tile in tiles]
    idx = pd.Index(tile_nums, name='tile')
    return xr.concat(tiles.values(), dim=idx).sortby('tile')


@delayed
def _median_no_dask(x, coarsening):
    n = len(x['xaxis_1'])
    return cubedsphere.block_median(x.chunk(), coarsening).compute()


@delayed
def upload_to_gcs(src, dest, save_op):
    logging.info("uploading %s to %s" % (src, dest))
    subprocess.check_call(['gsutil', '-q', 'cp',  src, dest])
    logging.info("uploading %s done" % dest)


@delayed
def run_all_delayeds(*args):
    pass


def output_names(key) -> dict:
    timestep, coarsenings = key
    urls = {}
    for coarsening in coarsenings:
        res = int(3072 // coarsening)
        category = 'sfc_data'
        output_file_name = f'gs://vcm-ml-data/2019-10-28-X-SHiELD-2019-10-05-multiresolution-extracted/coarsened/C{res}/{timestep}/{category}.nc'
        urls[coarsening] = output_file_name
    return urls


def coarsen_and_upload_surface(key):
    timestep, coarsenings= key
    output_file_names = output_names(key)

    ops = []

    # data downloading ops
    category = 'sfc_data'
    stored_resolution = 3702
    files = utils.file_names_for_time_step(timestep, category, resolution=stored_resolution)
    grouped_files = utils.group_file_names(files)
    opened = utils.map_ops(utils._open_remote_nc, grouped_files) 

    # coarse-graining
    with tempfile.TemporaryDirectory() as d:
        for coarsening in coarsenings:
            output_file_name = output_file_names[coarsening]
            logging.info("beggining processing job to %s" % output_file_name)
            coarse = utils.map_ops(_median_no_dask, opened, coarsening) 
            tiles = {key: combine_subtiles(val)
                    for key, val in coarse.items()}
            ds = concat_files(tiles)
            name = f"{d}/{coarsening}.name"
            save_op = ds.to_netcdf(name)
            upload_op = upload_to_gcs(name, output_file_name, save_op)
            ops.append(upload_op)

        all_ops = run_all_delayeds(*ops)
        #all_ops.compute(scheduler="single-threaded")
        all_ops.compute()

if __name__ == '__main__':
    import logging
    logging.basicConfig(level=logging.INFO)
    coarsen_and_upload_surface(timestep='20160805.114500', coarsenings=(8, 16, 32, 64))
