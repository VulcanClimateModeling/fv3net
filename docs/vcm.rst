VCM API
=======

.. currentmodule:: vcm

Computations
------------

.. autosummary::

   net_heating
   net_precipitation
   pressure_at_midpoint_log
   potential_temperature
   mass_integrate
   r2_score
   local_time
   coarsen_restarts_on_pressure
   coarsen_restarts_on_sigma
   
I/O
---

.. autosummary::

   open_tile
   open_delayed
   open_restarts
   open_restarts_with_time_coordinates

Cloud
-----

.. automodule:: vcm.cloud
   :members:


Full API
--------

.. automodule:: vcm
   :members:

.. automodule:: vcm.cubedsphere
   :members: