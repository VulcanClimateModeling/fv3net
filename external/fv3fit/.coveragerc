[run]
omit = tests/* setup.py fv3fit/sklearn/__main__.py fv3fit/keras/__main__.py
