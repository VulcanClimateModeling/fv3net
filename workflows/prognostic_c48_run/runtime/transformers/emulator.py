import dataclasses
from typing import Optional, Union, Sequence

import fv3fit.emulation.thermobasis.emulator
from fv3fit.emulation.thermobasis.xarray import get_xarray_emulator
from runtime.masking import get_mask, where_masked
from runtime.types import State

__all__ = ["Config", "Adapter"]


@dataclasses.dataclass
class Config:
    """
    Attributes:
        emulator: Either a path to a model to-be-loaded or an emulator
            configuration specifying model parameters.
        online: if True, the emulator will be replace fv3 physics for all
            humidities above level ``ignore_humidity_below``.
        train: if True, each timestep will be used to train the model.
        mask_kind: the type of mask. Defers to the functions
            ``runtime.masking.compute_{mask_kind}``. The default does not mask any
            of the emulator predictions.
        ignore_humidity_below: if mask_kind ="default", then use the fv3 physics
            instead of the emulator above this level.
    """

    emulator: Union[
        str, fv3fit.emulation.thermobasis.emulator.Config
    ] = dataclasses.field(default_factory=fv3fit.emulation.thermobasis.emulator.Config)
    # online parameters
    online: bool = False
    train: bool = True
    # will ignore the emulator for any z larger than this value
    # remember higher z is lower in the atmosphere hence "below"
    ignore_humidity_below: Optional[int] = None
    mask_kind: str = "default"


@dataclasses.dataclass
class Adapter:
    config: Config

    def __post_init__(self: "Adapter"):
        self.emulator = get_xarray_emulator(self.config.emulator)

    def predict(self, inputs: State) -> State:
        return self.emulator.predict(inputs)

    def apply(self, prediction: State, state: State):
        if self.config.online:
            updated_state = where_masked(
                state,
                prediction,
                compute_mask=get_mask(
                    self.config.mask_kind, self.config.ignore_humidity_below
                ),
            )
            state.update(updated_state)

    def partial_fit(self, inputs: State, state: State):
        if self.config.train:
            self.emulator.partial_fit(inputs, state)

    @property
    def input_variables(self) -> Sequence[str]:
        return self.emulator.input_variables
