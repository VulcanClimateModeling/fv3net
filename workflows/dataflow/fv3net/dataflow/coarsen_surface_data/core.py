import tempfile

import pandas as pd
import xarray as xr
from dask.delayed import delayed
from vcm import convenience, cubedsphere
import vcm

from gcs import upload_to_gcs

combine_subtiles = delayed(cubedsphere.combine_subtiles)


@delayed
def concat_files(tiles):
    tile_nums = [int(tile) for tile in tiles]
    idx = pd.Index(tile_nums, name="tile")
    return xr.concat(tiles.values(), dim=idx).sortby("tile")


@delayed
def _median_no_dask(x, coarsening):
    return cubedsphere.block_median(x.chunk(), coarsening).compute()


@delayed
def run_all_delayeds(*args):
    pass


def output_names(key) -> dict:
    timestep, coarsenings = key
    urls = {}
    for coarsening in coarsenings:
        output_file_name = (
            f"gs://vcm-ml-data/"
            "2019-10-28-X-SHiELD-2019-10-05-multiresolution-extracted/"
            "coarsened/C{res}/{timestep}/{category}.nc"
        )
        urls[coarsening] = output_file_name
    return urls


def coarsen_and_upload_surface(key):
    timestep, coarsenings = key
    output_file_names = output_names(key)

    ops = []

    # data downloading ops
    category = "sfc_data"
    stored_resolution = 3702
    files = convenience.file_names_for_time_step(
        timestep, category, resolution=stored_resolution
    )
    grouped_files = convenience.group_file_names(files)
    opened = convenience.map_ops(vcm.open_remote_nc, grouped_files)

    # coarse-graining
    with tempfile.TemporaryDirectory() as d:
        for coarsening in coarsenings:
            output_file_name = output_file_names[coarsening]
            logging.info("beggining processing job to %s" % output_file_name)
            coarse = convenience.map_ops(_median_no_dask, opened, coarsening)
            tiles = {key: combine_subtiles(val) for key, val in coarse.items()}
            ds = concat_files(tiles)
            name = f"{d}/{coarsening}.name"
            save_op = ds.to_netcdf(name)
            upload_op = upload_to_gcs(name, output_file_name, save_op)
            ops.append(upload_op)

        all_ops = run_all_delayeds(*ops)
        all_ops.compute(scheduler="single-threaded")


if __name__ == "__main__":
    import logging

    logging.basicConfig(level=logging.INFO)
    coarsen_and_upload_surface(timestep="20160805.114500", coarsenings=(8, 16, 32, 64))
