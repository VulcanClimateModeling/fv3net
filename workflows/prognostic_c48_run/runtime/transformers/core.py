import dataclasses
from typing import Iterable, Set, Protocol, Hashable

import vcm
from runtime.monitor import Monitor
from runtime.types import Diagnostics, State, Step


def strip_prefix(prefix: str, variables: Iterable[str]) -> Set[str]:
    return {k[len(prefix) :] for k in variables if k.startswith(prefix)}


class Predictor(Protocol):
    """Predictor interface for step transformers."""

    @property
    def input_variables(self) -> Iterable[Hashable]:
        """Variables needed as inputs for prediction."""
        pass

    def predict(self, inputs: State) -> State:
        """Given inputs return state predictions."""
        pass

    def apply(self, prediction: State, state: State):
        """Apply predictions to given state."""
        pass

    def partial_fit(self, inputs: State, state: State):
        """Do partial fit for online training."""
        pass


@dataclasses.dataclass
class StepTransformer:
    """Wrap a Step function with an ML prediction

    The wrapped function produces diagnostic outputs prefixed with
    ``self.label + "_"`` and trains/applies the ML model to ``state``
    depending on the user configuration.

    Attributes:
        model: A machine learning model that knows how to apply its updates.
        state: The mutable state being updated.
        label: Used for labeling diagnostic outputs and monitored tendencies.
        diagnostic_variables: The user-requested diagnostic variables.
        timestep: the model timestep in seconds.
    """

    model: Predictor
    state: State
    label: str
    diagnostic_variables: Set[str] = dataclasses.field(default_factory=set)
    timestep: float = 900

    @property
    def monitor(self) -> Monitor:
        return Monitor.from_variables(
            self.diagnostic_variables, self.state, self.timestep
        )

    @property
    def prefix(self) -> str:
        return self.label + "_"

    @property
    def inputs_to_save(self) -> Set[str]:
        return set(strip_prefix(self.prefix, self.diagnostic_variables))

    def transform(self, func: Step) -> Diagnostics:
        # do deep copy of DataArrays because func() typically modifies self.state
        inputs: State = {
            key: self.state[key].copy(deep=True) for key in self.model.input_variables
        }
        inputs_to_save: Diagnostics = {
            self.prefix + key: self.state[key] for key in self.inputs_to_save
        }

        before = self.monitor.checkpoint()
        diags = func()

        self.model.partial_fit(inputs, self.state)

        prediction = self.model.predict(inputs)
        change_due_to_prediction = self.monitor.compute_change(
            self.label, before, vcm.DerivedMapping({**before, **prediction})
        )
        self.model.apply(prediction, self.state)
        return {
            **diags,
            **inputs_to_save,
            **change_due_to_prediction,
        }

    def __call__(self, func: Step) -> Step:
        """Transform a function that updates the ``state``

        Similar to :py:class:`runtime.monitor.Monitor` but with ML prediction
        capability.

        Args:
            func: a function that updates the State and returns a dictionary of
                diagnostics (usually a method of :py:class:`runtime.loop.TimeLoop`).

        Returns:
            A function which calls ``func`` and optionally applies/trains an ML model
            in addition.
        """

        def step() -> Diagnostics:
            return self.transform(func)

        # functools.wraps modifies the type and breaks mypy type checking
        step.__name__ = func.__name__

        return step
