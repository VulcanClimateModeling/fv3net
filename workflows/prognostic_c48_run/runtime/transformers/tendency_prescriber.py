import dataclasses
import logging
from datetime import timedelta
from typing import Mapping, Set, Optional

import cftime
import xarray as xr

import pace.util
import loaders
import vcm
from runtime.monitor import Monitor
from runtime.types import Diagnostics, Step, State
from runtime.derived_state import DerivedFV3State
from runtime.conversions import quantity_state_to_dataset, dataset_to_quantity_state
from runtime.interpolate import time_interpolate_func, label_to_time

logger = logging.getLogger(__name__)


@dataclasses.dataclass
class TendencyPrescriberConfig:
    """Configuration for overriding tendencies from a step.
    
    Attributes:
        mapper_config: configuration of mapper used to load tendency data.
        variables: mapping from state name to name of corresponding tendency in
            provided mapper. For example: {"air_temperature": "fine_res_Q1"}.
    """

    mapper_config: loaders.MapperConfig
    variables: Mapping[str, str]
    reference_initial_time: Optional[str] = None
    reference_frequency_seconds: float = 900


@dataclasses.dataclass
class TendencyPrescriber:
    """Wrap a Step function and prescribe certain tendencies."""

    config: TendencyPrescriberConfig
    state: DerivedFV3State
    communicator: pace.util.CubedSphereCommunicator
    timestep: float
    diagnostic_variables: Set[str] = dataclasses.field(default_factory=set)

    def __post_init__(self: "TendencyPrescriber"):
        if self.communicator.rank == 0:
            logger.debug(f"Opening tendency override from: {self.config.mapper_config}")
        mapper = self.config.mapper_config.load_mapper()
        initial_label = self.config.reference_initial_time

        def mapper_func(time: cftime.DatetimeJulian) -> State:
            timestamp = vcm.encode_time(time)
            ds = mapper[timestamp]
            return {var: ds[var] for var in ds.data_vars}

        if initial_label is not None:
            self._mapper_func = time_interpolate_func(
                mapper_func,
                frequency=timedelta(seconds=self.config.reference_frequency_seconds),
                initial_time=label_to_time(initial_label),
            )
        else:
            self._mapper_func = mapper_func
        self._tendency_names = list(self.config.variables.values())

    def _open_tendencies_dataset(self, time: cftime.DatetimeJulian) -> xr.Dataset:
        tile = self.communicator.partitioner.tile_index(self.communicator.rank)
        if self.communicator.tile.rank == 0:
            ds = (
                xr.Dataset(self._mapper_func(time))
                .isel(tile=tile)[self._tendency_names]
                .load()
            )
        else:
            ds = xr.Dataset()
        tendencies = self.communicator.tile.scatter_state(dataset_to_quantity_state(ds))
        return quantity_state_to_dataset(tendencies)

    @property
    def monitor(self) -> Monitor:
        return Monitor.from_variables(
            self.diagnostic_variables, self.state, self.timestep
        )

    def _prescribe_tendency(self, func: Step) -> Diagnostics:
        tendencies = self._open_tendencies_dataset(self.state.time)
        before = self.monitor.checkpoint()
        diags = func()
        for variable_name, tendency_name in self.config.variables.items():
            with xr.set_options(keep_attrs=True):
                self.state[variable_name] = (
                    before[variable_name] + tendencies[tendency_name] * self.timestep
                )
        change_due_to_prescribing = self.monitor.compute_change(
            "tendency_prescriber", before, self.state
        )
        return {**diags, **change_due_to_prescribing}

    def __call__(self, func: Step) -> Step:
        """Override tendencies from a function that updates the State.
        
        Args:
            func: a function that updates the State and return Diagnostics.
            
        Returns:
            A function which calls ``func`` and prescribes a given change
            for specified variables.
        """

        def step() -> Diagnostics:
            return self._prescribe_tendency(func)

        step.__name__ = func.__name__
        return step
