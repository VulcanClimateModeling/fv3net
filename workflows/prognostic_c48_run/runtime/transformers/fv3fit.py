import dataclasses
from typing import Mapping, Iterable, Hashable, Sequence, Union

import xarray as xr
import fv3fit
from runtime.steppers.machine_learning import non_negative_sphum, MultiModelAdapter
from runtime.types import State
from runtime.names import SPHUM

__all__ = ["Config", "Adapter"]


@dataclasses.dataclass
class Config:
    """
    Attributes:
        url: Path to a model to-be-loaded, or sequence of paths.
        variables: Mapping from state names to name of corresponding tendency predicted
            by model. For example: {"air_temperature": "dQ1"}.
        limit_negative_humidity: if True, rescale tendencies to not allow specific
            humidity to become negative.
        online: if True, the ML predictions will be applied to model state.
    """

    url: Union[str, Sequence[str]]
    variables: Mapping[str, str]
    limit_negative_humidity: bool = True
    online: bool = True


@dataclasses.dataclass
class Adapter:
    config: Config
    timestep: float

    def __post_init__(self: "Adapter"):
        if isinstance(self.config.url, str):
            self.model = fv3fit.load(self.config.url)
        else:
            models = [fv3fit.load(url) for url in self.config.url]
            self.model = MultiModelAdapter(models)

    def predict(self, inputs: State) -> State:
        tendencies = self.model.predict(xr.Dataset(inputs))
        if self.config.limit_negative_humidity:
            dQ1, dQ2 = non_negative_sphum(
                inputs[SPHUM], tendencies["dQ1"], tendencies["dQ2"], self.timestep
            )
            tendencies.update({"dQ1": dQ1, "dQ2": dQ2})
        state_prediction: State = {}
        for variable_name, tendency_name in self.config.variables.items():
            with xr.set_options(keep_attrs=True):
                state_prediction[variable_name] = (
                    inputs[variable_name] + tendencies[tendency_name] * self.timestep
                )
        return state_prediction

    def apply(self, prediction: State, state: State):
        if self.config.online:
            state.update(prediction)

    def partial_fit(self, inputs: State, state: State):
        pass

    @property
    def input_variables(self) -> Iterable[Hashable]:
        return self.model.input_variables
