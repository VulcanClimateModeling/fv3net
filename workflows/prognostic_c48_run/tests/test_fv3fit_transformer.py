import joblib
import fv3fit
import vcm
import xarray as xr
from runtime.transformers.fv3fit import Config, Adapter
from runtime.transformers.core import StepTransformer
from machine_learning_mocks import get_mock_predictor


class MockDerivedState:
    def __init__(self, ds: xr.Dataset):
        self._mapper = ds

    def __getitem__(self, key: str) -> xr.DataArray:
        return vcm.DerivedMapping(self._mapper)[key]

    def __setitem__(self, key: str, value: xr.DataArray):
        self._mapper[key] = value

    def update(self, items):
        for key, value in items.items():
            self._mapper[key] = value


def regression_state(state, regtest):
    for v in sorted(state):
        print(v, joblib.hash(state[v].values), file=regtest)


def test_adapter_regression(state, regtest, tmpdir_factory):
    model_path = str(tmpdir_factory.mktemp("model"))
    mock = get_mock_predictor(dQ1_tendency=1 / 86400)
    fv3fit.dump(mock, model_path)

    adapted_model = Adapter(
        Config(
            model_path,
            {"air_temperature": "dQ1", "specific_humidity": "dQ2"},
            bias_correction_factor={"dQ1": -1},
        ),
        900,
    )
    transform = StepTransformer(
        adapted_model,
        MockDerivedState(state),
        "machine_learning",
        diagnostic_variables={
            "tendency_of_specific_humidity_due_to_machine_learning",
            "tendency_of_air_temperature_due_to_machine_learning",
            "tendency_of_internal_energy_due_to_machine_learning",
        },
        timestep=900,
    )

    def add_one_to_temperature():
        state["air_temperature"] += 1
        return {"some_diag": state["specific_humidity"]}

    out = transform(add_one_to_temperature)()

    # ensure tendency of internal energy is non-zero somewhere (GH#1433)
    max = abs(out["tendency_of_internal_energy_due_to_machine_learning"]).max()
    assert max.values.item() > 1e-6

    # sort to make the check deterministic
    regression_state(out, regtest)
