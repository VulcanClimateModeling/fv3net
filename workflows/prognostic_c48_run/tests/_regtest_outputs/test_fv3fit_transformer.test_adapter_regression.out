some_diag 8cbe81fbb51f1acc71f84e06b5fafdd4
storage_of_mass_due_to_machine_learning fcc46bebe36ea131688f8e15700e18d4
tendency_of_air_temperature_due_to_machine_learning a05dbea525b563eb5c0eaa9d10939bfc
tendency_of_internal_energy_due_to_machine_learning 99990b8e751d28a78dd2c3f4546ed97c
tendency_of_specific_humidity_due_to_machine_learning 0437557cc845e0f1b0f1bfdf0b59b17f
